src/workerd/io/hibernation-manager.c++

```diff
@@ -0,0 +1,235 @@
+  //
+  // We need to add the HibernatableWebSocket to each bucket in `tagToWs` corresponding to its tags.
+  //  1. Create the entry if it doesn't exist
+  //  2. Fill the TagListItem in the HibernatableWebSocket's tagItems array
+  size_t position = 0;
+  for (auto tag = tags.begin(); tag < tags.end(); tag++, position++) {
+    auto& tagCollection = tagToWs.findOrCreate(*tag, [&tag]() {
+      auto item = kj::heap<TagCollection>(
+          kj::mv(*tag), kj::heap<kj::List<TagListItem, &TagListItem::link>>());
+      return decltype(tagToWs)::Entry {
+          item->tag,
+          kj::mv(item)
+      };
+    });
+    // This TagListItem sits in the HibernatableWebSocket's tagItems array.
+    auto& tagListItem = refToHibernatable.tagItems[position];
+    tagListItem.hibWS = refToHibernatable;
+    tagListItem.tag = tagCollection->tag.asPtr();
+
+    auto& list = tagCollection->list;
+    list->add(tagListItem);
+    // We also give the TagListItem a reference to the list it was added to so the
+    // HibernatableWebSocket can quickly remove itself from the list without doing a lookup
+    // in `tagToWs`.
+    tagListItem.list = *list.get();
+  }
+
+  // Finally, we initiate the readloop for this HibernatableWebSocket.
+  kj::Promise<kj::Maybe<kj::Exception>> readLoopPromise = kj::evalNow([&] {
+    return readLoop(refToHibernatable);
+  }).then([]() -> kj::Maybe<kj::Exception> { return nullptr; },
+          [](kj::Exception&& e) -> kj::Maybe<kj::Exception> { return kj::mv(e); });
+
+  // Give the task to the HibernationManager so it lives long.
+  readLoopTasks.add(readLoopPromise.then(
+      [&refToHibernatable, this](kj::Maybe<kj::Exception>&& maybeError) -> kj::Promise<void> {
+    return handleSocketTermination(refToHibernatable, maybeError);
+  }));
+}
+
+kj::Vector<jsg::Ref<api::WebSocket>> HibernationManagerImpl::getWebSockets(
+    jsg::Lock& js,
+    kj::Maybe<kj::StringPtr> maybeTag) {
+  kj::Vector<jsg::Ref<api::WebSocket>> matches;
+  KJ_IF_MAYBE(tag, maybeTag) {
+    KJ_IF_MAYBE(item, tagToWs.find(*tag)) {
+      auto& list = *((*item)->list);
+      for (auto& entry: list) {
+        auto& hibWS = KJ_REQUIRE_NONNULL(entry.hibWS);
+        matches.add(hibWS.getActiveOrUnhibernate(js));
+      }
+    }
+  } else {
+    // Add all websockets!
+    for (auto& hibWS : allWs) {
+      matches.add(hibWS->getActiveOrUnhibernate(js));
+    }
+  }
+  return kj::mv(matches);
+}
+
+kj::Array<kj::byte> HibernationManagerImpl::serializeV8Value(
+    v8::Local<v8::Value> value,
+    v8::Isolate* isolate) {
+  jsg::Serializer serializer(isolate, jsg::Serializer::Options {
+    .version = 15,
+    .omitHeader = false,
+  });
+  serializer.write(value);
+  auto released = serializer.release();
+  JSG_REQUIRE(released.data.size() <= MAX_ATTACHMENT_SIZE, Error,
+      "A WebSocket 'attachment' cannot be larger than ",  MAX_ATTACHMENT_SIZE, " bytes");
```

Comment:
Since the user may not have visibility into how this is serialized, it would be helpful to include the `released.data.size()` in the error message so that they know by how much it was off.

Reply:
Moved this into `api::WebSocket` but the message has been updated https://github.com/cloudflare/workerd/pull/449/commits/610eda06f9dd7394d4b207e95d8322fa1883bf24#diff-6f0fba598574f6c0ada9d66bb6df723feee94392e3934530e125ef94b35fb9e6R344-R346